src/components/core/PivotTable.tsx

```diff
@@ -428,42 +432,90 @@ export class PivotTableInner extends BaseVisualization<IPivotTableInnerProps, IP
             };
-        }, initialValue);
+        }, this.resizedColumns);
     };
 
-    private autoresizeVisibleColumns = (columnApi: ColumnApi, previouslyResizedColumnIds: string[]) => {
+    private addToResizeColumn = (column: Column, source: ColumnEventSourceType): void => {
+        this.resizedColumns[this.getColumnIdentifier(column.getDefinition() as IGridHeader)] = {
+            width: column.getActualWidth(),
+            source,
+        };
+    };
+
+    private autoresizeVisibleColumns = async (
+        columnApi: ColumnApi,
+        previouslyResizedColumnIds: string[],
+        firstCall: boolean = true,
+    ): Promise<void> => {
+        const { execution } = this.state;
+
+        const tableIsNotScrolled = () => {
+            const horizontalPixelRange = this.gridApi.getHorizontalPixelRange();
+            const verticalPixelRange = this.gridApi.getVerticalPixelRange();
+            return horizontalPixelRange.left === 0 && verticalPixelRange.top === 0;
+        };
+
+        if (!(execution && tableIsNotScrolled())) {
+            return Promise.resolve();
+        }
+
+        if (!this.isColumnAutoresizeEnabled()) {
+            return Promise.resolve();
+        }
+
+        if (firstCall) {
+            await this.sleep(AGGRID_BEFORE_RESIZE_TIMEOUT);
+        }
+
         const displayedVirtualColumns = columnApi.getAllDisplayedVirtualColumns();
         const autoWidthColumnIds: string[] = this.getColumnIds(displayedVirtualColumns);
         if (previouslyResizedColumnIds.length >= autoWidthColumnIds.length) {
-            this.resizing = false;
-            this.resizedColumns = this.getResizedColumns(columnApi.getAllDisplayedVirtualColumns());
-            this.setState({
-                resized: true,
-            });
-            return;
+            this.resizedColumns = this.getResizedColumns(
+                columnApi.getAllDisplayedVirtualColumns(),
+                ColumnEventSourceType.AUTOSIZE_COLUMNS,
+            );
+            return Promise.resolve();
         }
 
-        const newColumnIds = difference(autoWidthColumnIds, previouslyResizedColumnIds);
-        this.autoresizeColumnsByColumnId(columnApi, newColumnIds);
-        setTimeout(() => {
-            this.autoresizeVisibleColumns(columnApi, autoWidthColumnIds);
-        }, AGGRID_RENDER_NEW_COLUMNS_TIMEOUT);
+        return new Promise(async resolve => {
+            const newColumnIds = difference(autoWidthColumnIds, previouslyResizedColumnIds);
+            this.autoresizeColumnsByColumnId(columnApi, newColumnIds);
+            await this.sleep(AGGRID_RENDER_NEW_COLUMNS_TIMEOUT);
+            resolve(this.autoresizeVisibleColumns(columnApi, autoWidthColumnIds, false));
+        });
     };
 
     private autoresizeColumnsByColumnId(columnApi: ColumnApi, columnIds: string[]) {
         columnApi.autoSizeColumns(columnIds);
     }
 
-    private autoresizeColumns = (
+    private sleep = async (delay: number): Promise<void> => {
```

Comment:
Could we put this sleep method to some "utils file" or something similar? I think that this function is general and you have the same function defined in test utils file. In case of change here you have to change the test sleep function as well.

Reply:
Done